src/vos/vos_dtx.c

```diff
@@ -932,47 +612,9 @@ vos_dtx_check_shares(struct umem_instance *umm, daos_handle_t coh,
 
 	D_ASSERT(intent == DAOS_INTENT_UPDATE);
 
-	/* Only OBJ record can be shared by new update. */
-	if (type != DTX_RT_OBJ) {
-		dtx_record_conflict(dth, dtx);
-
-		return dtx_inprogress(dtx, 5);
-	}
-
-	/* Here, if the obj/key has 'prepared' DTX, but current @dth is NULL,
-	 * then it is the race case between normal IO and rebuild: the normal
-	 * IO creates the obj/key before the rebuild request being handled.
-	 *
-	 * Under such case, we should (partially) commit the normal DTX with
-	 * the shared target (obj/key) to guarantee the rebuild can go ahead.
-	 */
-	if (dth == NULL) {
-		struct vos_container	*cont = vos_hdl2cont(coh);
-		int			 rc;
-
-		D_ASSERT(addr != NULL);
+	dtx_record_conflict(dth, dtx);
 
-		rc = vos_tx_begin(vos_cont2umm(cont));
-		if (rc == 0) {
-			umem_tx_add_ptr(umm, addr, sizeof(*addr));
-			*addr = UMOFF_NULL;
-			rc = vos_tx_end(vos_cont2umm(cont), 0);
-		}
-		if (rc != 0)
-			return rc;
-
-		return ALB_AVAILABLE_CLEAN;
-	}
-
-	D_ALLOC_PTR(dts);
-	if (dts == NULL)
-		return -DER_NOMEM;
-
-	dts->dts_type = type;
-	dts->dts_record = record;
-	d_list_add_tail(&dts->dts_link, &dth->dth_shares);
-
-	return ALB_AVAILABLE_CLEAN;
+	return dtx_inprogress(dtx, 5);
```

Comment:
This seems NOT suitable. Before new punch model introduced, we allow multiple concurrent modifications to update targets under the same object/dkey/akey. That is the DTX share mode. It should not cause any conflict.